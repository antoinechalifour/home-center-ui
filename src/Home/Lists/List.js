import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { graphql, compose } from 'react-apollo'
import gql from 'graphql-tag'
import DeleteIcon from 'react-icons/lib/md/clear'
import Card from 'ui/Card'
import Item from './Item'
import NewItem from './NewItem'

const Wrapper = styled(Card)`
  & + & {
    margin-top: 12px;
  }
`

const Name = styled.div`
  text-transform: uppercase;
  padding: 12px;
  text-align: center;
  box-shadow: 0 1px 3px rgba(0, 0, 0, .13);
  position: relative;

  svg {
    opacity: .54;
    cursor: pointer;
    position: absolute;
    right: 12px;
    top: 50%;
    transform: translateY(-50%);
  }
`

const List = ({ id, name, data, mutate }) => {
  const isLoading = data.loading

  if (isLoading) {
    return null
  }

  return (
    <Wrapper>
      <Name>
        <span>{name}</span>
        <DeleteIcon onClick={() => mutate({ variables: { id } })} />
      </Name>
      <ul>
        {data.list.items.map(x => <Item key={x.id} {...x} />)}
      </ul>
      <NewItem listId={id} />
    </Wrapper>
  )
}

List.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    list: PropTypes.shape({
      items: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          text: PropTypes.string.isRequired,
          done: PropTypes.bool.isRequired
        })
      ).isRequired
    })
  }).isRequired,
  mutate: PropTypes.func.isRequired
}

export default compose(
  graphql(
    gql`
      query ListQuery ($id: Int) {
        list (id: $id) {
          items {
            id,
            text,
            done
          }
        }
      }
  `,
    {
      options: ({ id }) => ({
        variables: { id }
      })
    }
  ),
  graphql(
    gql`
      mutation deleteList($id: Int) {
        deleteList(id: $id) {
          id
        }
      }
    `,
    {
      options: {
        refetchQueries: ['ListsQuery']
      }
    }
  )
)(List)
