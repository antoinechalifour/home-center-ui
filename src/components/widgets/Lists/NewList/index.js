import { graphql } from 'react-apollo'
import { createList } from 'queries/lists'
import NewList from './NewList'

const queryOptions = {}

export default graphql(createList, queryOptions)(NewList)
