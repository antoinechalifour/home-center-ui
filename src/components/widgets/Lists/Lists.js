import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import gqlLoaderHoc from 'components/GqlLoader'
import List from './List'

export function Lists ({ className, data }) {
  return (
    <Container>
      {data.lists &&
        <ListsWrapper>
          {data.lists.map(x => (
            <li key={x.id}>
              <List {...x} />
            </li>
          ))}
        </ListsWrapper>}
    </Container>
  )
}

Lists.propTypes = {
  data: PropTypes.shape({
    lists: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired
      })
    )
  }).isRequired
}

export default gqlLoaderHoc(Lists)

const Container = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
`

const ListsWrapper = styled.ul`
`
