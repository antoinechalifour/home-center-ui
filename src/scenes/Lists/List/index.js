import React, { Component } from 'react'
import { graphql, compose } from 'react-apollo'
import {
  getList,
  updateList,
  deleteList,
  listUpdated,
  listItemCreated,
  listItemUpdated,
  listItemDeleted
} from 'queries/lists'
import List from './List'

export class ListContainer extends Component {
  componentDidMount () {
    this.props.data.subscribeToMore({
      document: listUpdated,
      variables: {
        id: this.props.id
      },
      updateQuery: this.onListUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemCreated,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemCreated
    })

    this.props.data.subscribeToMore({
      document: listItemUpdated,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemUpdated
    })

    this.props.data.subscribeToMore({
      document: listItemDeleted,
      variables: {
        listId: this.props.id
      },
      updateQuery: this.onListItemDeleted
    })
  }

  onListUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updates = subscriptionData.data.listUpdated

    return {
      ...prev,
      list: {
        ...list,
        ...updates
      }
    }
  }

  onListItemCreated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const newItem = subscriptionData.data.listItemCreated

    return {
      ...prev,
      list: {
        ...list,
        items: [...list.items, newItem]
      }
    }
  }

  onListItemUpdated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const updatedItem = subscriptionData.data.listItemUpdated

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.map(x => (x.id === updatedItem.id ? updatedItem : x))
      }
    }
  }

  onListItemDeleted (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }

    const list = prev.list
    const deletedItem = subscriptionData.data.listItemDeleted

    return {
      ...prev,
      list: {
        ...list,
        items: list.items.filter(x => x.id !== deletedItem.id)
      }
    }
  }

  render () {
    return <List {...this.props} />
  }
}

const getListOptions = {
  options: ({ id }) => ({
    variables: { id }
  })
}

const updateListOptions = {
  name: 'updateList'
}

const deleteListOptions = {
  name: 'deleteList'
}

export default compose(
  graphql(getList, getListOptions),
  graphql(updateList, updateListOptions),
  graphql(deleteList, deleteListOptions)
)(ListContainer)
