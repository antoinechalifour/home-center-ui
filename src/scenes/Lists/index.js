import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import { getLists, listCreated, listDeleted } from 'queries/lists'
import Lists from './Lists'

export class ListsContainer extends Component {
  componentDidMount () {
    this.props.data.subscribeToMore({
      document: listCreated,
      updateQuery: this._onListCreated
    })

    this.props.data.subscribeToMore({
      document: listDeleted,
      updateQuery: this._onListDeleted
    })
  }

  _onListCreated (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }
    const lists = prev.lists
    const newList = subscriptionData.data.listCreated
    const isExistingList = lists.find(x => x.id === newList.id)

    if (isExistingList) {
      // If the list is already in the cache we ignore the event.
      return
    }

    // Otherwise we add it to the store
    return {
      ...prev,
      lists: [...lists, newList]
    }
  }

  _onListDeleted (prev, { subscriptionData }) {
    if (!subscriptionData.data) {
      return
    }
    const lists = prev.lists
    const deletedList = subscriptionData.data.listDeleted

    return {
      ...prev,
      lists: lists.filter(x => x.id !== deletedList.id)
    }
  }

  render () {
    return <Lists {...this.props} />
  }
}

export default graphql(getLists)(ListsContainer)
