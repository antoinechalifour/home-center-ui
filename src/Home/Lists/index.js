import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import Section, * as section from 'Home/Section'
import List from './List'
import NewList from './NewList'

const Content = styled(section.Content)`
  padding: 12px;
`

const FetchLists = ({ data }) => {
  return (
    <Section>
      <section.Title>
        My lists
      </section.Title>
      <Content>
        {data.loading && <section.Loader />}
        {data.lists && data.lists.map(x => <List key={x.id} {...x} />)}
        <NewList />
      </Content>
    </Section>
  )
}

FetchLists.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    lists: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired
      })
    )
  }).isRequired
}

export default graphql(
  gql`
  query ListsQuery {
    lists {
      id,
      name
    }
  }
`
)(FetchLists)
