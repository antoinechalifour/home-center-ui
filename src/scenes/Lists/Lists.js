import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Loader from 'ui/Loader'
import Title from 'ui/WidgetTitle'
import List from './List'
import NewList from './NewList'

export default function Lists ({ className, data }) {
  return (
    <div>
      <Title>My Lists</Title>
      {data.loading && <Loader />}
      {data.lists && <NewList />}
      {data.lists &&
        <ListsWrapper>
          {data.lists.map(x => (
            <li key={x.id}>
              <List {...x} />
            </li>
          ))}
        </ListsWrapper>}
    </div>
  )
}

Lists.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    lists: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired
      })
    )
  }).isRequired
}

const ListsWrapper = styled.ul`
  display: flex;
  padding: 12px 0;
  margin: auto;
  overflow-x: auto;

  > li {
    flex: 0 0 100%;

    + li {
      margin-left: 12px;
    }

    @media (min-width: 860px) {
      flex: 0 0 400px;
    }
  }

  > :first-child {
    margin-left: auto;
  }

  > :last-child {
    margin-right: auto;
  }

  > li + li {
    margin-left: 12px;
  }
`
