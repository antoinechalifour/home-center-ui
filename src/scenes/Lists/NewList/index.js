import { graphql } from 'react-apollo'
import { createList } from 'queries/lists'
import NewList from './NewList'

const queryOptions = {
  options: {
    refetchQueries: ['GetLists']
  }
}

export default graphql(createList, queryOptions)(NewList)
