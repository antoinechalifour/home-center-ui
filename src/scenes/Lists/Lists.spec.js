import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import Lists from './Lists'

const renderer = new ShallowRenderer()

describe('scenes/Lists', () => {
  it('Should render a loader when the data is fetching', () => {
    const props = {
      data: { loading: true }
    }
    const tree = renderer.render(<Lists {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should render a list of Lists otherwise', () => {
    const props = {
      data: {
        loading: false,
        lists: [
          {
            id: 1,
            name: 'List #1'
          },
          {
            id: 2,
            name: 'List #2'
          }
        ]
      }
    }
    const tree = renderer.render(<Lists {...props} />)

    expect(tree).toMatchSnapshot()
  })
})
