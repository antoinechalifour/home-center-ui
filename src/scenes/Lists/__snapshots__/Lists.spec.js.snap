// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes/Lists Should render a list of Lists otherwise 1`] = `
<div>
  <styled.h2>
    My Lists
  </styled.h2>
  <Apollo(NewList) />
  <styled.ul>
    <li>
      <Apollo(Apollo(Apollo(ListContainer)))
        id={1}
        name="List #1"
      />
    </li>
    <li>
      <Apollo(Apollo(Apollo(ListContainer)))
        id={2}
        name="List #2"
      />
    </li>
  </styled.ul>
</div>
`;

exports[`scenes/Lists Should render a loader when the data is fetching 1`] = `
<div>
  <styled.h2>
    My Lists
  </styled.h2>
  <Loader />
</div>
`;
